File: src/java/BrailleToText.java
import java.util.HashMap;
import java.util.Map;

public class BrailleToText { 

    private static final Map<String, String> BRAILLE_MAP = new HashMap<>();
    private static final Map<String, String> DIGIT_MAP = new HashMap<>();

    private static final String CAPITAL_INDICATOR = "⠠";   
    private static final String NUMBER_SIGN = "⠼";         
    private static final String HYPHEN = "⠤";             
    private static final String DOT_5 = "⠐";

    static {
        BRAILLE_MAP.put("⠁", "a"); BRAILLE_MAP.put("⠃", "b"); BRAILLE_MAP.put("⠉", "c");
        BRAILLE_MAP.put("⠙", "d"); BRAILLE_MAP.put("⠑", "e"); BRAILLE_MAP.put("⠋", "f");
        BRAILLE_MAP.put("⠛", "g"); BRAILLE_MAP.put("⠓", "h"); BRAILLE_MAP.put("⠊", "i");
        BRAILLE_MAP.put("⠚", "j"); BRAILLE_MAP.put("⠅", "k"); BRAILLE_MAP.put("⠇", "l");
        BRAILLE_MAP.put("⠍", "m"); BRAILLE_MAP.put("⠝", "n"); BRAILLE_MAP.put("⠕", "o");
        BRAILLE_MAP.put("⠏", "p"); BRAILLE_MAP.put("⠟", "q"); BRAILLE_MAP.put("⠗", "r");
        BRAILLE_MAP.put("⠎", "s"); BRAILLE_MAP.put("⠞", "t"); BRAILLE_MAP.put("⠥", "u");
        BRAILLE_MAP.put("⠧", "v"); BRAILLE_MAP.put("⠺", "w"); BRAILLE_MAP.put("⠭", "x");
        BRAILLE_MAP.put("⠽", "y"); BRAILLE_MAP.put("⠵", "z");

        BRAILLE_MAP.put(" ", " ");   
        BRAILLE_MAP.put("⠲", ".");   
        BRAILLE_MAP.put("⠂", ",");   
        BRAILLE_MAP.put("⠖", "!");   
        BRAILLE_MAP.put("⠦", "?");   
        BRAILLE_MAP.put("⠒", ":");   
        BRAILLE_MAP.put("⠆", ";");   
        BRAILLE_MAP.put("⠤", "-");   
        BRAILLE_MAP.put("⠄", "'");   
        BRAILLE_MAP.put("⠶", "\"");  
        BRAILLE_MAP.put("⠣", "(");   
        BRAILLE_MAP.put("⠜", ")");   

        DIGIT_MAP.put("⠁", "1"); DIGIT_MAP.put("⠃", "2"); DIGIT_MAP.put("⠉", "3");
        DIGIT_MAP.put("⠙", "4"); DIGIT_MAP.put("⠑", "5"); DIGIT_MAP.put("⠋", "6");
        DIGIT_MAP.put("⠛", "7"); DIGIT_MAP.put("⠓", "8"); DIGIT_MAP.put("⠊", "9");
        DIGIT_MAP.put("⠚", "0");
    }

    public static String translateUebGrade1ToText(String uebInput) {
        StringBuilder result = new StringBuilder();
        int i = 0;
        boolean inNumericMode = false;

        while (i < uebInput.length()) {
            String currentChar = String.valueOf(uebInput.charAt(i));

            // Capital Indicator
            if (currentChar.equals(CAPITAL_INDICATOR)) {
                if (i + 1 < uebInput.length()) {
                    String nextBraille = String.valueOf(uebInput.charAt(i + 1));
                    String mappedText = BRAILLE_MAP.get(nextBraille);
                    if (mappedText != null && Character.isLetter(mappedText.charAt(0))) {
                        result.append(mappedText.toUpperCase());
                        i += 2;
                        continue;
                    }
                }
                i += 1;
                continue;
            }

            // Dot 5 (Parentheses Handling)
            if (currentChar.equals(DOT_5)) {
                if (i + 1 < uebInput.length()) {
                    String nextBraille = String.valueOf(uebInput.charAt(i + 1));
                    if (nextBraille.equals("⠣")) {
                        result.append("(");
                        i += 2;
                        continue;
                    }
                    if (nextBraille.equals("⠜")) {
                        result.append(")");
                        i += 2;
                        continue;
                    }
                }
                i += 1;
                continue;
            }

            // Numeric Indicator
            if (currentChar.equals(NUMBER_SIGN)) {
                inNumericMode = true;
                i += 1;
                continue;
            }

            String textChar = BRAILLE_MAP.get(currentChar);
            
            if (textChar != null) {
                if (inNumericMode) {
                    if (DIGIT_MAP.containsKey(currentChar)) {
                        result.append(DIGIT_MAP.get(currentChar));
                    } 
                    else if (currentChar.equals(HYPHEN)) {
                        // Hyphen terminates numeric mode
                        inNumericMode = false;
                        result.append('-');
                    } 
                    else if (currentChar.equals("⠲")) { // Period / Decimal
                        // Lookahead check
                        boolean isDecimal = false;
                        if (i + 1 < uebInput.length()) {
                            String nextC = String.valueOf(uebInput.charAt(i + 1));
                            if (DIGIT_MAP.containsKey(nextC)) {
                                isDecimal = true;
                            }
                        }
                        result.append(".");
                        if (!isDecimal) {
                            inNumericMode = false;
                        }
                    }
                    else if (currentChar.equals("⠂")) {
                        // Comma maintains numeric mode
                        result.append(",");
                    }
                    else {
                        inNumericMode = false;
                        result.append(textChar);
                    }
                } else {
                    result.append(textChar);
                }
            }
            i += 1;
        }

        return result.toString().trim();
    }
    
    public static void main(String[] args) {
         System.out.println(BrailleToText.translateUebGrade1ToText("⠼⠁⠃⠉⠲⠙⠑")); 
    }
}

File: src/javascript/braille-to-text.js
function translateUebGrade1ToText(uebInput) {
    const BRAILLE_MAP = {
        '⠁': 'a', '⠃': 'b', '⠉': 'c', '⠙': 'd', '⠑': 'e', '⠋': 'f',
        '⠛': 'g', '⠓': 'h', '⠊': 'i', '⠚': 'j', '⠅': 'k', '⠇': 'l',
        '⠍': 'm', '⠝': 'n', '⠕': 'o', '⠏': 'p', '⠟': 'q', '⠗': 'r',
        '⠎': 's', '⠞': 't', '⠥': 'u', '⠧': 'v', '⠺': 'w', '⠭': 'x',
        '⠽': 'y', '⠵': 'z',
        ' ': ' ', '⠲': '.', '⠂': ',', '⠖': '!', '⠦': '?',
        '⠒': ':', '⠆': ';', '⠤': '-', '⠄': "'", '⠶': '"',
        '⠣': '(', '⠜': ')'
    };

    const CAPITAL_INDICATOR = '⠠';   
    const NUMBER_SIGN = '⠼';  
    const DOT_5 = '⠐';       

    const DIGIT_MAP = {
        '⠁': '1', '⠃': '2', '⠉': '3', '⠙': '4', '⠑': '5',
        '⠋': '6', '⠛': '7', '⠓': '8', '⠊': '9', '⠚': '0',
    };

    let result = [];
    let i = 0;
    let inNumericMode = false;

    while (i < uebInput.length) {
        const char = uebInput[i];

        // Capital Indicator
        if (char === CAPITAL_INDICATOR) {
            if (i + 1 < uebInput.length) {
                const nextBraille = uebInput[i + 1];
                if (BRAILLE_MAP.hasOwnProperty(nextBraille) && /[a-z]/.test(BRAILLE_MAP[nextBraille])) {
                    result.push(BRAILLE_MAP[nextBraille].toUpperCase());
                    i += 2;
                    continue;
                }
            }
            i += 1;
            continue;
        }

        // Dot 5 (Parentheses)
        if (char === DOT_5) {
             if (i + 1 < uebInput.length) {
                 const nextBraille = uebInput[i + 1];
                 if (nextBraille === '⠣') {
                     result.push('(');
                     i += 2; 
                     continue;
                 }
                 if (nextBraille === '⠜') {
                     result.push(')');
                     i += 2;
                     continue;
                 }
             }
             i += 1;
             continue;
        }

        // Numeric Indicator
        if (char === NUMBER_SIGN) {
            inNumericMode = true;
            i += 1;
            continue;
        }

        if (BRAILLE_MAP.hasOwnProperty(char)) {
            const textChar = BRAILLE_MAP[char];

            if (inNumericMode) {
                if (DIGIT_MAP.hasOwnProperty(char)) {
                    result.push(DIGIT_MAP[char]);
                } else if (char === '⠤') {
                    // Hyphen terminates numeric mode
                    inNumericMode = false;
                    result.push('-');
                } else if (char === '⠲') {
                    // Decimal Lookahead
                    let isDecimal = false;
                    if (i + 1 < uebInput.length) {
                        if (DIGIT_MAP.hasOwnProperty(uebInput[i + 1])) {
                            isDecimal = true;
                        }
                    }
                    result.push('.');
                    if (!isDecimal) {
                        inNumericMode = false;
                    }
                } else if (char === '⠂') {
                    // Comma continues numeric mode
                    result.push(textChar);
                } else {
                    inNumericMode = false;
                    result.push(textChar);
                }
            } else {
                result.push(textChar);
            }
        }
        i += 1;
    }

    return result.join('').trim();
}